plugins/catalyst/lib/Conversion/MQTOptToCatalystQuantum/MQTOptToCatalystQuantum.cpp

```diff
@@ -0,0 +1,1198 @@
+    auto pi34 = rewriter.create<MulFOp>(op.getLoc(), pi4, three);
+
+    // Compute φ - π/2
+    auto phiMinusPi2 = rewriter.create<SubFOp>(op.getLoc(), phi, pi2);
+    // Compute λ - π/2
+    auto lambdaMinusPi2 = rewriter.create<SubFOp>(op.getLoc(), lambda, pi2);
+
+    // RZ(λ − π/2)
+    auto rz1 = rewriter.create<catalyst::quantum::CustomOp>(
+        op.getLoc(), outQubitTypes, TypeRange{}, ValueRange{lambdaMinusPi2},
+        inQubitsValues, "RZ", false, inCtrlQubits, ValueRange{});
+
+    // RX(π/2)
+    auto rx1 = rewriter.create<catalyst::quantum::CustomOp>(
+        op.getLoc(), outQubitTypes, TypeRange{}, ValueRange{pi2},
+        rz1.getResults(), "RX", false, inCtrlQubits, ValueRange{});
+
+    // RZ(3/4 π)
+    auto rz2 = rewriter.create<catalyst::quantum::CustomOp>(
+        op.getLoc(), outQubitTypes, TypeRange{}, ValueRange{pi34},
+        rx1.getResults(), "RZ", false, inCtrlQubits, ValueRange{});
+
+    // RX(π/2)
+    auto rx2 = rewriter.create<catalyst::quantum::CustomOp>(
+        op.getLoc(), outQubitTypes, TypeRange{}, ValueRange{pi2},
+        rz2.getResults(), "RX", false, inCtrlQubits, ValueRange{});
+
+    // RZ(φ − π/2)
+    auto rz3 = rewriter.create<catalyst::quantum::CustomOp>(
+        op.getLoc(), outQubitTypes, TypeRange{}, ValueRange{phiMinusPi2},
+        rx2.getResults(), "RZ", false, inCtrlQubits, ValueRange{});
+
+    // Replace the original U gate with the decomposed sequence
+    rewriter.replaceOp(op, rz3.getResults());
+    return success();
+  }
+};
+
+template <>
+struct ConvertMQTOptSimpleGate<opt::PeresOp> final
+    : OpConversionPattern<opt::PeresOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(opt::PeresOp op, opt::PeresOp::Adaptor adaptor,
+                  ConversionPatternRewriter& rewriter) const override {
+    // Extract operands
+    auto inQubitsValues = adaptor.getInQubits();
+    const auto posCtrlQubitsValues = adaptor.getPosCtrlInQubits();
+    const auto negCtrlQubitsValues = adaptor.getNegCtrlInQubits();
+
+    SmallVector<Value> inCtrlQubits(posCtrlQubitsValues.begin(),
+                                    posCtrlQubitsValues.end());
+    inCtrlQubits.append(negCtrlQubitsValues.begin(), negCtrlQubitsValues.end());
+
+    // Output type setup
+    const Type qubitType =
+        catalyst::quantum::QubitType::get(rewriter.getContext());
+    const SmallVector<Type> qubitTypes(
+        inQubitsValues.size() + inCtrlQubits.size(), qubitType);
+    auto outQubitTypes = TypeRange(qubitTypes);
+
+    // Peres = Toffoli(q2,q1,q0) CNOT(q2,q1)
+    auto ccnot1 = rewriter.create<catalyst::quantum::CustomOp>(
+        op.getLoc(), outQubitTypes, TypeRange{}, ValueRange{}, inQubitsValues,
+        "Toffoli", false, inCtrlQubits, ValueRange{});
+
+    auto cnot2 = rewriter.create<catalyst::quantum::CustomOp>(
+        op.getLoc(), outQubitTypes, TypeRange{}, ValueRange{},
+        ccnot1.getResults(), "CNOT", false, inCtrlQubits, ValueRange{});
```

Comment:
This is not correct. The Peres Gate in mqt-core is a two-target gate that decomposes down to a CNOT and an X gate. It does not use 3 qubits. Same holds for the inverse.